### PR TITLE
Only collect mono items from reachable blocks

### LINKED ISSUE #85836: dead-code optimize `if const { expr }` even in opt-level=0
`inline_const` tracking issue: https://github.com/rust-lang/rust/issues/76001

Inspired by @oli-obk's comment in https://github.com/rust-lang/rust/pull/85828#discussion_r642046617

For things like `if const { expr } { 0 } else { do something complex }` it'd be nice to not have to even give LLVM the code for the else branch to have to emit.  Ought to save time and binary size.

Tentatively marked mir-opt, since we presumably still need to make the MIR for it, in order to borrow-check and such the unreachable code.  But might need to be codegen-time instead to pick up on monomorphized inline constants.

Hopefully should be doable relatively easily when the const is an inline const, since that forces const folding already, and thus there should just be a `const true` in the MIR.

### PR DESCRIPTION
**Pull request by saethlin**
Fixes the wrong comment pointed out in: https://github.com/rust-lang/rust/pull/121421#discussion_r1537378431
Moves the analysis to use the worklist strategy: https://github.com/rust-lang/rust/pull/121421#discussion_r1501840823
Also fixes https://github.com/rust-lang/rust/issues/85836, using the same reachability analysis

### CODE REVIEW

**Comment by saethlin** [comment]
Great. There is actually measurable query overhead here.

**Comment by saethlin** [comment]
I accidentally removed the optimization from mono item collection in the last perf run. Oops.

@bors try @rust-timer queue

**Comment by saethlin** [comment]
r? cjgillot

**Comment by saethlin** [comment]
This seems like a fiddly invariant to uphold in cg_clif. But also, this probably improves compile time especially of large programs when using cranelift. So maybe this is still a good idea. :shrug: @bjorn3 what do you think?

**Comment by bjorn3** [comment]
The cg_clif change seems fine to me.

**Comment by pnkfelix** [comment]
* fixed an important bug.
* from the comment history, it looks like the minor restricted regressions were anticipated.
* marking as triaged.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_middle/src/mir/traversal.rs [hunk 1]
```diff
@@ -279,3 +279,97 @@ pub fn reverse_postorder<'a, 'tcx>(
...
+
+impl<'a, 'tcx> Iterator for MonoReachable<'a, 'tcx> {
+    type Item = (BasicBlock, &'a BasicBlockData<'tcx>);
+
+    fn next(&mut self) -> Option<(BasicBlock, &'a BasicBlockData<'tcx>)> {
+        while let Some(idx) = self.worklist.iter().next() {
+            self.worklist.remove(idx);
+            if !self.visited.insert(idx) {
+                continue;
+            }
```
Comment (riking) [inline compiler/rustc_middle/src/mir/traversal.rs:357]: This is quadratic for the case where all blocks reference all earlier blocks, right?

Comment (saethlin) [reply compiler/rustc_middle/src/mir/traversal.rs:357]: Could be. Would that be remarkable?

**Pull request merged**